dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java

```diff
@@ -418,6 +424,27 @@ private void bulkUpsertWorkflows(User authUser) {
         }
     }
 
+    private void checkToolTokens(User authUser, Long userId, String organization) {
+        List<Token> tokens = tokenDAO.findByUserId(userId);
+        List<Tool> tools = userContainers(authUser, userId);
+        if (organization != null && !organization.isEmpty()) {
+            tools.removeIf(tool -> !tool.getNamespace().equals(organization));
+        }
+        Token gitLabToken = Helper.extractToken(tokens, TokenType.GITLAB_COM.toString());
+        Token quayioToken = Helper.extractToken(tokens, TokenType.QUAY_IO.toString());
+        Set<Registry> uniqueRegistry = new HashSet<>();
+        tools.forEach(tool -> uniqueRegistry.add(tool.getRegistry()));
+        if (uniqueRegistry.size() == 0 && quayioToken == null) {
+            throw new CustomWebApplicationException("Please add a Quay.io token", HttpStatus.SC_BAD_REQUEST);
+        }
+        if (uniqueRegistry.contains(Registry.QUAY_IO) && quayioToken == null) {
+            throw new CustomWebApplicationException("Have Quay.io tools but no Quay.io token", HttpStatus.SC_BAD_REQUEST);
```

Comment:
Elaborate (i.e. why is this situation a bad thing?)

Reply:
Also thought it was agreed upon that if the user actually has Quay tools, but then unlinked their token, and then refresh all...rather than silently succeeding, there would be a notification.

Reply:
Same here